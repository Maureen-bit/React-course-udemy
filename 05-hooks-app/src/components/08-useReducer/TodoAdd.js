import React from 'react'
import { useForm } from '../../hooks/useForm';

export const TodoAdd = ({handleAddTodo}) => {
    /*COMPONENTE CREADO PARA SEPARAR LA LÓGICA DEL FORMULARIO DEL COMPONENTE PRINCIPAL */

        /* Valores que retorna el custom hook creado para el manejo de los cambios en los inputs y los valores
    del formulario */
    const [ { description }, handleInputChange, reset ] = useForm({
        /*tiene que ser la misma key que le puse al input */
        description: ""
    });

    const handleSubmit = (event) => {
        event.preventDefault();

        if (description.trim().length <= 1 ) {
            return;
        };

        /* Nueva tarea que se enviará al reducer para que se modifique el estado y agregar la nueva tarea */
        const newTodo = {
            id: new Date().getTime(),
            desc: description,
            done: false
        }

        handleAddTodo(newTodo);
        reset();
    }

  return (
   <>
        <h4>Agregar tarea</h4>
        <hr/>

        <form onSubmit={handleSubmit}>
            <input  
                type="text"
                name="description"
                className="form-control"
                placeholder="Aprender..."
                autoComplete="off"
                onChange={handleInputChange}
                value={description}
            />
            <button type='submit' className='btn btn-outline-primary mt-1 btn-block'>Agregar</button>
        </form>
   </>
  )
}
