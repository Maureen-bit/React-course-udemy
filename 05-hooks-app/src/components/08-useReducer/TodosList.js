import React from 'react'
import { TodoListItem } from './TodoListItem'

export const TodosList = ({todos,  handleToggle, handleDelete}) => {

  return (
    <ol className='list-group list-group-flush'>
            {
                todos.map((todo, index) => (
                    <TodoListItem 
                        key={todo.id}
                        todo={todo}
                        index={index}
                        handleToggle={handleToggle}
                        handleDelete={handleDelete}
                    /> )
                )
            }
        </ol>
  )
}
