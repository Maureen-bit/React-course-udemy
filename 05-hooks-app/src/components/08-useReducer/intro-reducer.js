/* Establecemos el estado inicial */
const initialState = [{
    id: 1,
    todo: 'Comprar pan',
    done: false
}]

/* Creación del reducer */
const todoReducer  = (state = initialState, action) => {
    /* la accion es la que modifica el state */
    if (action?.type === 'agregar') {
        /* No se puede usar push porque ese modifica el objeto inicial entonces se destructura todo el arreglo y al final se agrega el
        nuevo objeto */
        return [...state, action.payload]
    }

    return state;
};

let todos = todoReducer();

const newTodo = {
    id: 3,
    todo: 'Comprar huevos',
    done: false
}

/* Creación de la acción con los parámetros estándar, el tipo de acción: lo que se va a hacer (agregar, modificar o eliminar) y
payload: el nuevo estado */
const agregarTodoAction = {
    type: 'agregar',
    payload: newTodo
};

todos = todoReducer( todos, agregarTodoAction );

console.log("🚀 ~ file: intro-reducer.js ~ line 17 ~ todos", todos)