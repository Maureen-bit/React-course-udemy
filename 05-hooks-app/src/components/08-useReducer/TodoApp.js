import React, { useReducer, useEffect } from 'react'
import './styles.css';
import { TodoAdd } from './TodoAdd';
import todoReducer from './todoReducer';
import { TodosList } from './TodosList';

export const TodoApp = () => {

    const init = () => {
        /* Establece el estado inicial del reducer */
        return JSON.parse(localStorage.getItem('todos')) || []
    };

    const [ todos, dispatch ] = useReducer(todoReducer, [], init);
    /* en destructuración de arreglos se puede cambiar el nombre de las variables ya que el resultado
    es posicionado, es decir, state es posición cero asi se llame todos.
    DISPATCH: función que recibe una action para ejecutar  y enviar la información al reducer específico que 
    se está usando ya que es pasado por referencia y cuando cambie el state va a redibujar lo que cambie */



    useEffect(() => {
      localStorage.setItem('todos', JSON.stringify(todos) )
    }, [todos])

    const handleDelete = (todoId) => {
    const action = {
        type: 'DELETE',
        payload: todoId
    };
    dispatch(action);
    };

    const handleToggle = (todoId) => {
        const action = {
            type: 'TOGGLE',
            payload: todoId
        };

        dispatch(action);
    }

    const handleAddTodo = (newTodo) =>{
        dispatch({
            type: 'ADD',
            payload: newTodo
        }) 
    }
    
  return (
    <>
    <h1>TodoApp ({ todos.length })</h1>
    <hr />
    <div className='row'>
        <div className='col-7'>
            <TodosList
                todos={todos}
                handleToggle={handleToggle}
                handleDelete={handleDelete}
             />
        </div>
        <div className='col-5'>
            <TodoAdd 
                handleAddTodo={handleAddTodo}
            />
        </div>
    </div>
    </>
  )
}
