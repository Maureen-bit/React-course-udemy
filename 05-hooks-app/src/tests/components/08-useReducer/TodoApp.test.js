import React from 'react';
import { shallow, mount } from 'enzyme';
import { TodoApp } from '../../../components/08-useReducer/TodoApp';
import { demoTodos } from '../../dataTest/demoTodos';
import { act } from '@testing-library/react';

describe('Tests in <TodoApp /> component', () => { 
    const wrapper = shallow( <TodoApp/> );
    Storage.prototype.setItem = jest.fn(() => {});

    test('it should renders correctly', () => { 
        expect(wrapper).toMatchSnapshot();
     })

     /* This test is not working because an issue about compatibility with this version of some test libraries with 
     React. Mount line is breaking out the test */
     test('it should add a new TODO', () => { 
         /* mount: monta todo el componente y los hijos del componente además de renderizarlo completamente, 
         si tiene componentes hijos también se renderizaran*/
         const wrapper = mount(<TodoApp/>);
         act(() => {
            wrapper.find('TodoAdd').prop('handleAddTodo')(demoTodos[0])
            wrapper.find('TodoAdd').prop('handleAddTodo')(demoTodos[1])
         })
        expect(wrapper.find('h1').text().trim()).toBe('TodoApp (2)');
        expect(localStorage.setItem).toHaveBeenCalledTimes(2);
      });

      test('it should delete a todo', () => { 
        wrapper.find('TodoAdd').prop('handleAddTodo')(demoTodos[0]);
        wrapper.find('TodosList').prop('handleDelete')(demoTodos[0].id);
        expect(wrapper.find('h1').text().trim()).toBe('TodoApp (0)');
       });
 });
