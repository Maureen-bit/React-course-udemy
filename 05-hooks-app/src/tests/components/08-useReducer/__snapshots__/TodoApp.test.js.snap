// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in <TodoApp /> component it should renders correctly 1`] = `
<Fragment>
  <h1>
    TodoApp (
    0
    )
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-7"
    >
      <TodosList
        handleDelete={[Function]}
        handleToggle={[Function]}
        todos={Array []}
      />
    </div>
    <div
      className="col-5"
    >
      <TodoAdd
        handleAddTodo={[Function]}
      />
    </div>
  </div>
</Fragment>
`;
